### PR TITLE
Factor out a DenseMatrixBase implementation header

### PR DESCRIPTION
**Pull request by dschwen**
Refs #2089 (there might be more cases where this is required to enable external apps to create new template instantiations - for example with dual numbers)

### CODE REVIEW

**Comment by dschwen** [comment]
Darn, that header doesn't get installed. I must have missed something...

**Comment by roystgnr** [comment]
Did you run include/rebuild_include_HEADERS.sh?

**Comment by roystgnr** [comment]
Ah, looks like you got it now.

**Comment by dschwen** [comment]
Uhm, no, but I edited libmesh/include/include_HEADERS

**Comment by dschwen** [comment]
Let me make teh change using that script you mentioned to keep the diffs clean going forward

**Comment by roystgnr** [comment]
Ha, I was wondering how there ended up being a white space change in the diff!  Your hand-written changes were better than the script's version...

**Comment by dschwen** [comment]
As an aside, will there ever be a unity build for libmesh... :-) ?

**Comment by dschwen** [comment]
ARGH!!! This is STILL not getting installed!!

**Comment by jwpeterson** [comment]
@dschwen, in case you did not see it? https://github.com/libMesh/libmesh/wiki/Adding-or-removing-source-files

**Comment by dschwen** [comment]
@jwpeterson that link helps. I ran 

```
cd include && ./rebuild_include_HEADERS.sh
cd ..
cd src && ./rebuild_libmesh_SOURCES.sh
cd ..
cd include/libmesh && ./rebuild_makefile.sh
cd -
./bootstrap
```

Bit the git diff showed no changes, so I guess I should be fine. Yet the Test/InstallCheck is failing (and I broke MOOSE test?! -  `petsc.h` not found?!).

What. Up?

**Comment by jwpeterson** [comment]
> What. Up?

There is an issue with CIVET today, I believe @milljm and @permcody are working on that.

**Comment by dschwen** [comment]
DenseMatrix needs some `auto` or `T` in place of `Real` to enable templating on DualNumbers

**Comment by permcody** [comment]
New package release on CIVET today. I'm helping @milljm iron out some of the issues that arise when there are competing packages. We have merged in a new change that should help with stability in this situation. To be clear, we haven't deployed the package officially yet with MOOSE, but it's available to CIVET and being picked up _before_ the repositories are cloned/cleaned (e.g. during pre-check).

**Comment by jwpeterson** [comment]
> DenseMatrix needs some auto or T in place of Real to enable templating on DualNumbers

@lindsayad did some of this already I thought?

**Comment by dschwen** [comment]
Thinks like `Real DenseMatrix::min()` are biting me in the behind here. I want the return type to be 

```
  auto min () const -> decltype(libmesh_real((*this)(0,0)));
```

But that doesn't work because `definition of 'libMesh::DenseMatrix<double>' is not complete until the closing '}'`

**Comment by dschwen** [comment]
This seems to work:

```
auto DenseMatrix<T>::min () const -> decltype(libmesh_real(T(0)))
```

**Comment by jwpeterson** [comment]
OK... I guess I don't get why the `min` should be real-valued if T==std::complex

**Comment by dschwen** [comment]
Yeah, me neither, but I'm not eager to break behavior that somebody may rely on.

**Comment by jwpeterson** [comment]
It's been like that since 2006.... be1f873f3dec1fbd92dc45a3b27f13172de58f61, the comments suggest  @roystgnr wanted it that way at the time... he may have been thinking in terms of matrix norms I guess.

**Comment by dschwen** [comment]
Ok, builds now and allows me to use DenseMatrix<DualReal> in my app.

**Comment by roystgnr** [comment]
I'd guess that my only thought was that having min() return min(real()) was likely to be more useful than having it return NaN or throw an error.  But in hindsight "more useful" is also "less safe" so I wouldn't object if someone wanted to change that behavior.

File: include/numerics/dense_matrix.h [hunk 1]
```diff
@@ -1015,17 +1015,17 @@ Real DenseMatrix<T>::l1_norm () const
...
 template<typename T>
 inline
-Real DenseMatrix<T>::linfty_norm () const
+auto DenseMatrix<T>::linfty_norm () const -> decltype(std::abs(T(0)))
 {
   libmesh_assert (this->_m);
   libmesh_assert (this->_n);
 
-  Real rowsum = 0.;
+  decltype(std::abs(T(0))) rowsum = 0.;
```
Comment (jwpeterson) [inline include/numerics/dense_matrix.h]: Just curious, does `auto rowsum = std::abs(T(0));` work to initialize this, or do you really need `decltype` here?

Comment (dschwen) [reply include/numerics/dense_matrix.h]: Yeah, I'm pretty sure that would work. It looks obvious now, but I went through so many iteration with the decltype arguments that initially I didn't think of that. I'll make the change.

**Comment by lindsayad** [comment]
@dschwen How did you get this working in your app? I'm getting undefined references with respect to `DenseMatrix::_multiply_blas` upon instantiation of `DenseMatrix<ADReal>`

**Comment by lindsayad** [comment]
I’m thinking of including the impl files, throwing in some empty voids for the blas lapack private methods, and then explicitly instantiating on ADReal in a MOOSE .C file to get this done. That should work right?

**Comment by lindsayad** [comment]
Well with error messages instead of empty voids...and will have to make sure to manually set the use_blas_lapack flag to false.

**Comment by dschwen** [comment]
If you have Marmot access check the `deep_neural_network_318` branch on my fork.

**Comment by dschwen** [comment]
Also https://github.com/idaholab/moose/pull/13294. I figured I might have to add stubs for all the blas crap. But so far I have not run into this issue, and I have already successfully run simulations using  `DenseMatrix<DualReal>`...

**Comment by lindsayad** [comment]
I am guessing that this can only be on a Mac which does lazy symbol binding...? I get undefined references when linking on linux for the blas/lapack stuff

**Pull request merged**